packages/babel-core/src/config/caching-a.js

```diff
@@ -0,0 +1,108 @@
+// @flow
+
+import gensync, { type Handler, type Gensync } from "gensync";
+import { CacheConfigurator } from "./caching";
+
+export type SimpleCacheConfigurator = SimpleCacheConfiguratorFn &
+  SimpleCacheConfiguratorObj;
+
+type SimpleCacheConfiguratorFn = {
+  (boolean): void,
+  <T>(handler: () => T): T,
+};
+type SimpleCacheConfiguratorObj = {
+  forever: () => void,
+  never: () => void,
+  using: <T>(handler: () => T) => T,
+  invalidate: <T>(handler: () => T) => T,
+};
+
+type CacheEntry<ResultT, SideChannel> = Array<{
+  value: ResultT,
+  valid: SideChannel => boolean,
+}>;
+
+export type { CacheConfigurator };
+
+/**
+ * Given a function with a single argument, cache its results based on its argument and how it
+ * configures its caching behavior. Cached values are stored strongly.
+ */
+export function makeStrongCache<ArgT, ResultT, SideChannel>(
+  handler: (ArgT, CacheConfigurator<SideChannel>) => Handler<ResultT>,
+): Gensync<[ArgT, SideChannel], ResultT> {
+  return makeCachedFunction<ArgT, ResultT, SideChannel, *>(new Map(), handler);
+}
+
+/**
+ * Given a function with a single argument, cache its results based on its argument and how it
+ * configures its caching behavior. Cached values are stored weakly and the function argument must be
+ * an object type.
+ */
+export function makeWeakCache<
+  ArgT: {} | Array<*> | $ReadOnlyArray<*>,
+  ResultT,
+  SideChannel,
+>(
+  handler: (ArgT, CacheConfigurator<SideChannel>) => Handler<ResultT>,
+): Gensync<[ArgT, SideChannel], ResultT> {
+  return makeCachedFunction<ArgT, ResultT, SideChannel, *>(
+    new WeakMap(),
+    handler,
+  );
+}
+
+type CacheMap<ArgT, ResultT, SideChannel> =
+  | Map<ArgT, CacheEntry<ResultT, SideChannel>>
+  | WeakMap<ArgT, CacheEntry<ResultT, SideChannel>>;
+
+function makeCachedFunction<
+  ArgT,
+  ResultT,
+  SideChannel,
+  // $FlowIssue https://github.com/facebook/flow/issues/4528
+  Cache: CacheMap<ArgT, ResultT, SideChannel>,
+>(
+  callCache: Cache,
+  handler: (ArgT, CacheConfigurator<SideChannel>) => Handler<ResultT>,
+): Gensync<[ArgT, SideChannel], ResultT> {
+  return gensync(function* cachedFunction(arg: ArgT, data: SideChannel) {
+    let cachedValue: CacheEntry<ResultT, SideChannel> | void = callCache.get(
+      arg,
+    );
+
+    if (cachedValue) {
+      for (const { value, valid } of cachedValue) {
+        if (valid(data)) return value;
```

Comment:
Do you think we should consider allowing this to be async too? We're currently stuck with the one `statSync` because it is sync-only. We shouldn't expose gensync to consumers, but we could repeat the `.then` check here too and allow returning a promise, if we wanted.

Not needed now and easy to expand later, but curious for your thoughts.

Reply:
I didn't change it because I still have to finish refactoring the caching logic, but I will do it :+1:

Reply:
Cool!

Reply:
# TODO